### ISSUE TITLE
Add method for creating PipeReader from IMessageReader<ReadOnlySequence<byte>>

### ISSUE THREAD

**Issue by davidfowl** [davidfowl/BedrockFramework#43]
Some protocols take a large stream of bytes and introduce framing to produce a smaller stream of bytes. In some cases those frames don't matter and we want to read them as another stream of bytes. An example of this are chunked response encoding in HTTP/1.1. Each incoming frame is wrapped in a length prefix (in hex) and a new line. Most frameworks ignore this framing and present the user with a stream of bytes. 

To make this pattern easier to implement, we should make it possible to take a `PipeReader` and an `IMessageReader<ReadOnlySequence<byte>>` and return a new `PipeReader` that implements that generic de-framing and buffering.

The current prototype is https://github.com/davidfowl/BedrockFramework/compare/davidfowl/pipe-reader

**Comment by YairHalberstadt** [comment]
I'm happy to give this a go. Can you explain where the prototype is lacking please?

**Comment by davidfowl** [comment]
The prototype currently doesn't handle any of the buffering semantics of a `PipeReader` and the unit test specified shows one of the worst cases. But generally, when data is being buffered, all new read data needs to be buffered. For example:

https://github.com/davidfowl/BedrockFramework/compare/davidfowl/pipe-reader#diff-622cb55f718f1568563332ba39c71b89R39-R56

This logic is incorrect, but based on the amount of data buffered, if there's a backlog then trim it here (it's what lead me file this issue https://github.com/dotnet/runtime/issues/1248). Something like this (pseudo code):

```C#
var consumedBytes = _message.Slice(_message.Start, consumed).Length;

// Removes from the front of the buffer
_backlog.Consume(consumedBytes);

// Append any unconsumed bytes to the end of the buffer
var unconsumed = _message.Slice(consumed);

foreach (var m in unconsumed)
{
    _backlog.Write(m.Span);
}
```

If there's data in the backlog when you're doing the next read then it needs to be appended to the current backlog. 

https://github.com/davidfowl/BedrockFramework/compare/davidfowl/pipe-reader#diff-622cb55f718f1568563332ba39c71b89R86-R94

Then the other things that need to be implemented:

- TryRead - Like ReadAsync but non blocking
- Cancellation, it needs to be tested
- Completion, you can't read or write after completing
- It needs performance tests. We'll likely have to implement memory pooling around the backlog. Today it's allocating non-pooled arrays on append.

**Comment by YairHalberstadt** [comment]
Would you recommend copying ArrayBufferWriter and implementing that API our-self? It looks like it should be simple enough to do.

**Comment by davidfowl** [comment]
Yep, That makes sense. Implement our own and then add pooling to it.

**Comment by YairHalberstadt** [comment]
I had a chance to implement ConsumableArrayBufferWriter today.

I'm now trying to integrate it with your prototype.

However I'm having problems with this line:

https://github.com/davidfowl/BedrockFramework/compare/davidfowl/pipe-reader#diff-622cb55f718f1568563332ba39c71b89R93

We return a ReadOnlySequence wrapping the backlog.

When the user calls `AdvanceTo` he passes in a `SequencePosition` from the backlog, not from the `ReadResult.Buffer` returned by the underlying `_reader`. In order to call `AdvanceTo` on the underlying pipe reader we have to have a `SequencePosition` from the `ReadResult.Buffer`. How do we map from a `SequencePosition` on the `backlog` to a `SequencePosition` in the `ReadResult.Buffer`.

Also why do we use a `ArrayBufferWriter`? Why not just store the `_messages` directly?

Thanks for your help. This is my first experience with System.IO.Pipelines, so it might take me a while to grok!

**Comment by YairHalberstadt** [comment]
> You should be assigning _message to the new ReadOnlySequence like below.

> `_message = new ReadOnlySequence<byte>(_backlog.WrittenMemory)`

But it's not the correct sequence for the underlying pipereader.

> What do you mean? We need to copy the buffer since AdvanceTo since the user may not have consumed the entire buffer.

But why copy the bytes? Why not store a list of ReadOnlySequences, or combine the ReadOnlySequences into one?

**Comment by YairHalberstadt** [comment]
Ok, I think I've managed to wrap my head around whats happening here.

Thanks!

**Comment by davidfowl** [comment]
> But why copy the bytes? Why not store a list of ReadOnlySequences, or combine the ReadOnlySequences into one?

Because that will likely be extremely inefficient.

**Comment by YairHalberstadt** [comment]
Initial PR at https://github.com/davidfowl/BedrockFramework/pull/53

**Comment by davidfowl** [comment]
Fixed in #53
